### PR TITLE
fix(distillation): spurious negative kl_loss from bf16 reductions in the chunked path

### PR DESCRIPTION
**Pull request by opooladz**
# PR: fix(distillation): spurious negative `kl_loss` from bf16 reductions in the chunked path

**Base:** `vnext` · **Head:** `fix/chunked-distill-kl-bf16-precision` (2 commits + regression tests)

## Problem

With the chunked distillation loss (`logits_chunk_size > 0`) and bf16 `compute_dtype`, the logged
`kl_loss` of a pure forward-KL run can read **negative** — impossible for a true KL
(Gibbs' inequality) — and carries an O(2.5e-2) error floor at large vocabularies, making small KL
values uninterpretable. The `distill_xent_loss` / `teacher_entropy_loss` diagnostics are worse:
at vocab ~2.5e5 with peaked teachers (logit magnitudes ~50–70), `distill_xent_loss` can read
**−1.4 for a true value of +0.97**.

Minimal demonstration (V = 248,320, bf16, *bit-identical* student and teacher logits, so true
KL ≡ 0):

```
per-token KL formula, eager:              +0.00000   (exact)
same formula, under jax.jit:              +0.02480
chunked path end-to-end (jit):            +0.02478
same, fp32 inputs:                        +0.00000   (exact)
```

Reproduced across 6 PRNG seeds: +0.0227 … +0.0255, never 0. The sign of the offset is
backend/fusion/data dependent — on TPU it swings negative per batch, producing the negative
readings whenever the true KL is below the floor.

## Root cause

Under `jax.jit`, the teacher and student log-sum-exp reductions lower to **separate fused reduce
ops** (their surrounding graphs differ: `keepdims` reshape, consumer sets, AD/remat structure of
the differentiated student branch vs the `stop_gradient` teacher branch). Each fused reduction
independently chooses its accumulation tiling. A bf16 accumulation over 2.5e5 elements has
order-dependent rounding of O(2.5e-2) at this scale; two differently-tiled summations of the same
data round **differently**, so the error does not cancel in `lse_t − lse_s` — even on identical
inputs. The xent/entropy defect is related but distinct: teacher probabilities normalized with the
bf16-rounded lse sum to `1 + O(1e-2)` instead of 1, leaking `O(δ·|logits|)` into the O(1)
readings.

## Fix (commit 1)

- Compute the per-token KL as `Σ p_t (t − s) − (lse_t − lse_s)` with **both logsumexp reductions
  accumulated in fp32**: `jax.nn.logsumexp(x.astype(float32))`. The upcast is a producer fused
  into the reduce — **no fp32 `[..., V]` buffer is materialized**, so the memory profile of the
  chunked path is unchanged. The KL sum is carried in fp32 through the chunk scan.
- Normalize the metric-only xent/entropy diagnostics with fp32 probabilities inside their
  reductions.
- **The training gradient is unchanged**: teacher-side terms are `stop_gradient`, so
  `∂loss/∂student` has the same dependency structure as before (verified equal to the analytic
  KD gradient `(p_s − p_t)/N` within bf16-softmax tolerance, and unchanged vs. the prior code).

Verification at V = 248,320, bf16, jit:

| case | before | after |
|---|---|---|
| identical inputs (true KL = 0), 6 seeds | +0.0227…+0.0255 | **0.0 exactly** |
| true KL ≈ 0.008 | reads ~0.033 | 0.0073…0.0089 |
| true KL = 0.362 | +0.385 | +0.359 |
| `distill_xent` (true 0.974) | −1.375 | +0.9727 |
| `teacher_entropy` (true 0.612) | 0.586 | 0.6133 |
| `∇` vs analytic | 1.73e-3 | 1.70e-3 |

After the fix the identity `distill_xent − teacher_entropy ≈ kl_loss` holds per step, giving a
live cross-check of the metric pipeline.

## Opt-in top-1 agreement metric (commit 2)

`EASYDEL_LOG_TOP1_AGREEMENT=1` adds `top1_agreement` — the masked fraction of tokens where
`argmax(student) == argmax(teacher)` — accumulated through the chunk scan. An argmax comparison
involves no near-equal subtraction, so it is insensitive to the entire error class above and
complements `kl_loss` as a hard agreement signal. Two extra vocab reductions over logits already
in registers (fused; no buffer); the flag is read at import time, so the disabled branch compiles
to nothing and **default behavior is unchanged**.

## Tests

`tests/trainers/test_distillation_kl_precision.py`:
- `test_kl_exactly_zero_for_identical_logits_under_jit` — the invariance that catches this whole
  error class (pre-fix code fails it at large vocab);
- kl vs fp32-reference tracking at two noise levels (+ non-negativity);
- xent/entropy vs reference + the `xent − entropy ≈ kl` identity;
- gradient vs analytic KD gradient.

Existing `tests/trainers/test_distillation_loss_math.py` passes unchanged (public signatures
untouched).

### CODE REVIEW

**Comment by opooladz** [comment]
u already took care of the same bug in a later commit
